api/service.go

```diff
@@ -514,6 +525,199 @@ func (s *service) RemoveAppToken(ctx context.Context, req *pb.RemoveAppTokenRequ
 	return &pb.RemoveAppTokenReply{}, nil
 }
 
+func (s *service) AddFile(server pb.API_AddFileServer) error {
+	log.Debugf("received add file request")
+
+	scope, ok := server.Context().Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+
+	req, err := server.Recv()
+	if err != nil {
+		return err
+	}
+	var name, projID string
+	switch payload := req.GetPayload().(type) {
+	case *pb.AddFileRequest_Header_:
+		projID = payload.Header.ProjectID
+		name = payload.Header.Name
+	default:
+		return fmt.Errorf("project ID is required")
+	}
+
+	proj, err := s.getProjectForScope(server.Context(), projID, scope)
+	if err != nil {
+		return err
+	}
+
+	sendEvent := func(event *pb.AddFileReply_Event) {
+		if err := server.Send(&pb.AddFileReply{
+			Payload: &pb.AddFileReply_Event_{
+				Event: event,
+			},
+		}); err != nil {
+			log.Errorf("error sending event: %v", err)
+		}
+	}
+
+	sendErr := func(err error) {
+		if err := server.Send(&pb.AddFileReply{
+			Payload: &pb.AddFileReply_Error{
+				Error: err.Error(),
+			},
+		}); err != nil {
+			log.Errorf("error sending error: %v", err)
+		}
+	}
+
+	reader, writer := io.Pipe()
+	waitCh := make(chan struct{})
+	go func() {
+		defer close(waitCh)
+		for {
+			req, err := server.Recv()
+			if err == io.EOF {
+				_ = writer.Close()
+				return
+			} else if err != nil {
+				_ = writer.CloseWithError(err)
+				sendErr(err)
+				return
+			}
+			switch payload := req.GetPayload().(type) {
+			case *pb.AddFileRequest_Chunk:
+				if _, err := writer.Write(payload.Chunk); err != nil {
+					sendErr(err)
+					return
+				}
+			default:
+				sendErr(fmt.Errorf("invalid request"))
+				return
+			}
+		}
+	}()
+
+	var size string
+	eventCh := make(chan interface{})
+	defer close(eventCh)
+	go func() {
+		for e := range eventCh {
+			event, ok := e.(*iface.AddEvent)
+			if !ok {
+				log.Error("unexpected event type")
+				continue
+			}
+			if event.Path == nil { // This is a progress event
+				sendEvent(&pb.AddFileReply_Event{
+					Name:  event.Name,
+					Bytes: event.Bytes,
+				})
+			} else {
+				size = event.Size // Store size for use in the final response
+			}
+		}
+	}()
+
+	pth, err := s.ipfsClient.Unixfs().Add(
+		server.Context(),
+		ipfsfiles.NewReaderFile(reader),
+		options.Unixfs.Pin(true),
+		options.Unixfs.Progress(true),
+		options.Unixfs.Events(eventCh))
+	if err != nil {
+		return err
+	}
+
+	_, err = s.collections.Files.Create(server.Context(), pth, name, proj.ID)
+	if err != nil && !strings.Contains(err.Error(), store.ErrUniqueExists.Error()) {
+		return err
+	}
+
+	sendEvent(&pb.AddFileReply_Event{
```

Comment:
Then, send the last event with the final path.

Reply:
`sendEvent` will only log if an error happened while sending. That sounds good for progress events since they aren't a critical thing. But I wonder what might happen if this last event fails on sending. Apart from logging it, this `sendEvent` assumes all goes well. 

Also, the `s.collections.Files.Create` was already called at this point, so that could be a problem?

Considering not missing anything; possible solution: 
- make `sendEvent` return an `error`. While sending progress, ignore it. Not ignore it on this last send. 
- Switch first sending event, and then `.Create()` call? Not sure about this last point since maybe both orderings are problematic in both ways, so just mentioning it if you have a more clear picture if one of them might be better than the other.

Extra question: if last `sendEvent` fails, the channel with receiver will close anyway and they'll understand that something bad happened since was closed before receiving the last result, right? (just checking they won't block forever waiting for it).

Reply:
>     * make `sendEvent` return an `error`. While sending progress, ignore it. Not ignore it on this last send.

Good call! 

>     * Switch first sending event, and then `.Create()` call? Not sure about this last point since maybe both orderings are problematic in both ways, so just mentioning it if you have a more clear picture if one of them might be better than the other.

Right yeah, I was wondering the same about the order of things. Not ideal in either case. Though, the remote uses the event with final `Path` to indicate to the client that they can stop receiving. In this setup, the `sendEvent` must come last, otherwise the client would have already left the room by time the remote calls `Create`, ignoring any errors that happen there. 

> Extra question: if last `sendEvent` fails, the channel with receiver will close anyway and they'll understand that something bad happened since was closed before receiving the last result, right? (just checking they won't block forever waiting for it).

In this case, (currently) the remote would end as normal, causing an EOF on the client. This sends to its wait channel, which is used to wrap things up. However, we can catch the error as mentioned above, in which case it would be used to wrap things up on the client.

Reply:
SGTM